modules/guides/src/GuidesStorage.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace Drupal\guides;
+
+/**
+ * Copyright (C) 2019 PRONOVIX GROUP BVBA.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+ * USA.
+ */
+
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Link;
+use Drupal\Core\Site\Settings;
+use Drupal\guides\Exception\FileNotFoundException;
+
+/**
+ * Defines GuidesStorage for Guides files, links and directories.
+ */
+final class GuidesStorage implements GuidesStorageInterface {
+
+  /**
+   * The settings of the site.
+   *
+   * @var \Drupal\Core\Site\Settings
+   */
+  protected $settings;
+
+  /**
+   * The cache backend service.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cacheBackend;
+
+  /**
+   * GuidesStorage constructor.
+   *
+   * @param \Drupal\Core\Site\Settings $settings
+   *   The settings of the site.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cache backend.
+   */
+  public function __construct(Settings $settings, CacheBackendInterface $cache_backend) {
+    $this->settings = $settings;
+    $this->cacheBackend = $cache_backend;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  private function getDirectory(): string {
+    return DRUPAL_ROOT . ($this->settings->get('guides_dir') ?? '/guides');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFiles(): array {
+    $directory = $this->getDirectory();
+    $cid = 'guides_files:' . $directory;
+    $data_cached = $this->cacheBackend->get($cid);
+
+    if ($data_cached) {
+      $guides = $data_cached->data;
+    }
+    else {
+      $iterator = new \RecursiveDirectoryIterator($directory, \FilesystemIterator::CURRENT_AS_FILEINFO | \FilesystemIterator::SKIP_DOTS);
+      $filter = new \RecursiveCallbackFilterIterator($iterator, function ($current, $key, $iterator) {
+        if ($iterator->hasChildren()) {
+          return TRUE;
+        }
+        $path = $current->getPathname();
+        if (is_file($path)) {
+          return substr($current->getFilename(), -3) == '.md';
```

Comment:
Is `$current->getPathinfo()->getExtension()` available? (Or something similar) I want you to retrieve the extension from the pathinfo.